### ISSUE TITLE
ibrier measure (still) flawed

### ISSUE THREAD

**Issue by HerrMo** [mlr-org/mlr#2426]
1.
Stumbeld over the same error #2425 last week and was about to open an issue. I think the second FIXME concerned the indexing of the `crps` output. As long as `reference = TRUE` the first slot of `crps` is reserved for the Kaplan-Meier-estimate and the returned number was - indeed - wrong (as it was the KM-estimate regradless which model was fit). So now the FIXME is obsolete.
But I think it is now going to throw an error as the output of `crps` is no longer a 2-dim object. `crps(perror, times = max.time)[1L]` should do.

Also: I think it is not necessary to explicitly compute the survival probabilities via `probs = predictSurvProb(getLearnerModel(model, more.unwrap = TRUE), newdata = newdata, times = grid)` as `pec` handles that for you as long as there is a `predSurvProb` method for the learner. 

2.
Additionally there might be another error when predicting evaluation data.

If you have for example a setup like this

`mod <- train(lrn, task, subset = train)`
`pred <- predict(mod, task, subset = test)`

I think the performance via the ibrier measure will be computed w.r.t. the training data and not to the test data. Since in 
https://github.com/mlr-org/mlr/blob/e9e80dfdfbf8685c47ba116081144d159d1d5d7e/R/measures.R#L1426 
`model$subset` is referring to the training data (or do I miss something?). 
Shouldn't it be for example something like 
`newdata = getTaskData(task)[-model$subset, ]`?

**Comment by PhilippPro** [comment]
I will have a look at it. Ideally I think we should have an own function for that and not rely on the pec package anymore, but this is some work...

**Comment by PhilippPro** [comment]
I think the problem is, that pec requires the whole `newdata` matrix from the new data to predict probabilities for each time point, but the only thing that is available in mlr is the `pred` object which only contains the predicted and actual survival time. So either we change the pred object and make it possible to predict probabilities or we make it possible to pass the "newdata" to the performance function.

**Comment by PhilippPro** [comment]
@mllg can you give a comment on that?

**Comment by IyarLin** [comment]
@PhilippPro What do you mean by "predicted survival times" in "only thing that is available in mlr is the pred object which only contains the predicted and actual survival time". I'm new to the mlr internals and couldn't find the piece of code that produces the `pred` object.

**Comment by PhilippPro** [comment]
i just mean code like this: 

```
cox_lrn <- makeLearner(cl = "surv.coxph")
lung = lung %>% mutate(status = as.logical(status)) %>% filter(complete.cases(.))
task_surv = makeSurvTask(data = lung, target = c("time", "status"))
task = task_surv
model = train(cox_lrn, task)
pred = predict(model, task)
```
If you look at `pred$data` you will find what I described. 

I rethought a bit this weekend and think that it would be "cleaner" and more to the standards of mlr if the probabilities would be in the prediction object. I will try to make a PR.

**Comment by IyarLin** [comment]
I would suggest predicting at every **train set** unique death time point. This is because the prediction can't change between those times so it's the most compact way of representing these predictions. I would also recommend estimating the errors at every unique **test set** death time point. That is because the error measurement can't change between these points. 

I'm not very familiar with the mlr internals yet so I'll give an example using non mlr code and hopefully it won't be too hard to implement. If it turns out to be a hassle let me know and I'll try making a PR myself.
```

library(survival); library(pec); library(dplyr)
data(lung)
lung = lung %>% 
  mutate(status = as.logical(ifelse(status == 2, 0, 1))) %>% 
  filter(complete.cases(.))

train_ind <- sample.int(nrow(lung), nrow(lung)*0.75)
train_data <- lung[train_ind, ]
test_data <- lung[-train_ind, ]

model = coxph(Surv(time, status) ~ ., data = train_data, x = T) # lookout for the x = T argument, recently it's became crucial to the predictSurvProb function

if(class(model) %in% c("coxph", "rfsrc")){
  unique_death_times <- c(0, sort(unique(train_data$time[train_data$status == 1])))
  surv_prob <- predictSurvProb(object = model, 
                               newdata = test_data, 
                               times = unique_death_times)
  dimnames(surv_prob) <- list(1:nrow(test_data), 
                              unique_death_times) # assign this matrix to a slot within pred
  
  }

## calculate ibrier:
grid = sort(unique(test_data$time[test_data$status == 1])) # this should be the grid object in the new ibrier measure
prob_columns = sapply(grid, function(t) max(which(t >= unique_death_times)))
probs = surv_prob[, prob_columns] # this should be the probs object in the new ibrier measure
perror = pec(object = probs, formula = Surv(time, status) ~ 1, 
             data = test_data, times = grid, exact = F, start = min(grid), reference = F) # this should be the perror object in the new ibrier measure

crps(perror, times = max(grid))[1L]
```

**Comment by PhilippPro** [comment]
I created a branch where I am currently working: [](https://github.com/mlr-org/mlr/tree/survival_probabilities_ibrier)

Not done completely yet, but what works is this:

```
cox_lrn <- makeLearner(cl = "surv.coxph", predict.type = "prob")
lung = lung %>% mutate(status = as.logical(status)) %>% filter(complete.cases(.))
task_surv = makeSurvTask(data = lung, target = c("time", "status"))
task = task_surv
model = train(cox_lrn, task)
pred = predict(model, task)
getPredictionResponse(pred)
getPredictionProbabilities(pred)
performance(pred, measures = list(cindex.uno), model = model, task = task)
newdata = getTaskData(task)
performance(pred, measures = list(ibrier2), task = task, newdata = newdata)
```

Resampling does not work, have to work on it tomorrow.

**Comment by IyarLin** [comment]
> Why are these two statements true? Is this valid for every existing survival method?
I didn't find it in the literature, but for the survival estimate at time t to change you need to observe a death, otherwise there's nothing to suggest there's something different at that point. But I might be wrong.

**Comment by PhilippPro** [comment]
I looked at two examples with coxph and rfsrc and it seems to be true...

I updated the [branch](https://github.com/mlr-org/mlr/tree/survival_probabilities_ibrier), resampling is working now. I changed some of the internals and did not check a lot, so it is not done yet but it works. The time points for probabilities are now the train and test time points, this was necessary for resampling and later putting the predictions of the resampling together (and I think it is also better to have a prediction of each time point of train and test data.)
Feel free to test.

**Comment by IyarLin** [comment]
Sorry for not getting around to test out the code, having problems with mlr on my laptop lately. Just a small comment though: When you run the following code:

`lung = lung %>% mutate(status = as.logical(status))`

You're actually setting the status to be TRUE as can be seen by running 

`table(lung$status)`

You should instead run:

`lung = lung %>% mutate(status = status == 2)`

(looking on the help page for lung you can see that 2 stands for dead, 1 for censorship)

**Comment by PhilippPro** [comment]
Hi,
thanks for the comment. I currently do not use this test code anymore, as I have a pull request where there is also a test included. See here: https://github.com/mlr-org/mlr/pull/2430

You can already install, use and test the branch by executing 
`install_github("mlr-org/mlr", ref = "survival_probabilities_ibrier")`

Feel free to comment or suggest improvements.
